### DISCUSSION TITLE
Dialog Discussion

### DISCUSSION THREAD

**Discussion by danirabbit** [elementary/stylesheet#592] [Ideas]
Exploring some dialog design stuff for Greenfield. 

A huge consideration with Greenfield is contrast. So you'll notice that buttons stand out a lot more here. Overall, superfluous details are made much more subtle, but I don't feel like overall things are "flatter". I think we're using depth a lot more here actually.

Here's a comparison:

## BEFORE
![Screenshot from 2020-03-09 10 52 58@2x](https://user-images.githubusercontent.com/7277719/76242587-335f0080-61f4-11ea-98c9-7177b430b5f0.png)

## AFTER
![Screenshot from 2020-03-09 09 57 58@2x](https://user-images.githubusercontent.com/7277719/76241252-ed08a200-61f1-11ea-8cc2-0752a4b59f7d.png)

## ACCENT COLORS
One of the considerations we're working with is customizable accent colors. This is something that developers can set to reinforce their own branding, but we also want users to be able to set their own default accent color. The idea here is that offering some customization options can satisfy the need that drives user to change their stylesheet, but in a way that doesn't lead to application breakage.

You'll notice the following screenshots don't have a shadow in that action box. That's just because I did this before messing with that style. Consider these all just sketches. I haven't checked anything for contrast and everything is subject to change.

![Screenshot from 2020-03-09 09 44 08@2x](https://user-images.githubusercontent.com/7277719/76241312-feea4500-61f1-11ea-9422-6feb1b099216.png)
![Screenshot from 2020-03-09 09 44 31@2x](https://user-images.githubusercontent.com/7277719/76241314-ff82db80-61f1-11ea-9754-3ebd231852cc.png)
![Screenshot from 2020-03-09 09 44 52@2x](https://user-images.githubusercontent.com/7277719/76241318-001b7200-61f2-11ea-9679-a362862b8bf1.png)
![Screenshot from 2020-03-09 09 45 21@2x](https://user-images.githubusercontent.com/7277719/76241319-00b40880-61f2-11ea-93ca-d852c399835b.png)

This obviously leaves us in a place where we need to differentiate between red accent color and destructive action. Using a lighter button style leans well into this:

![Screenshot from 2020-03-09 09 43 21@2x](https://user-images.githubusercontent.com/7277719/76241307-fd208180-61f1-11ea-8ef0-b31408b989f2.png)

![Screenshot from 2020-03-09 09 28 25@2x](https://user-images.githubusercontent.com/7277719/76241303-fc87eb00-61f1-11ea-917b-844db83d0ade.png)

But it brings up questions about if this works as well in the dark style:

![Screenshot from 2020-03-09 10 31 36@2x](https://user-images.githubusercontent.com/7277719/76241221-deba8600-61f1-11ea-9bc2-36f8f0965277.png)
![Screenshot from 2020-03-09 10 33 52@2x](https://user-images.githubusercontent.com/7277719/76241225-dfebb300-61f1-11ea-908e-d69f9d87979d.png)

I made this branch: https://github.com/elementary/granite/pull/374

And here's the CSS I've been playing with in Inspector:

```
dialog {
    background: @base_color;
    border-radius: 0 0 7px 7px;
    box-shadow:
        inset 1px 0 0 0 alpha (@bg_highlight_color, 0.2),
        inset -1px 0 0 0 alpha (@bg_highlight_color, 0.2),
        inset 0 -1px 0 0 alpha (@bg_highlight_color, 0.3);
}

dialog headerbar.default-decoration {
    background: @base_color;
    border-radius: 7px 7px 0 0;
}

dialog decoration {
    border-radius: 6px;
    box-shadow:
        0 0 0 1px alpha (#000, 0.1),
        0 8px 10px 1px alpha (#000, 0.14),
        0 3px 14px 2px alpha (#000, 0.12),
        0 5px 5px -3px alpha (#000, 0.4),
        0 14px 28px transparent;
}

dialog .dialog-action-area {
    margin: 7px;
}

dialog .dialog-action-box {
    background-color: alpha(#000, 0.03);
    background-image:
        linear-gradient(
            to bottom,
            alpha (
                #000,
                0.15
            ) 1px,
            alpha (
                #000,
                0.08
            ) 1px,
            alpha (
                #000,
                0
            ) 3px
        );
    border-top: 1px solid alpha (@bg_highlight_color, 0.3);
    margin-top: 24px;
}

dialog .dialog-vbox {
    margin: -2px;
}

dialog .primary {
    //color: @accent_color_900;
    font-size: 1.22em;
    font-weight: 500;
}

dialog button {
    border: none;
    border-radius: 3px;
    background-color: @base_color;
    background-image:
        linear-gradient(
            to bottom,
            alpha(@bg_highlight_color, 0.2),
            alpha(@bg_highlight_color, 0.0)
        );
    box-shadow:
        inset 1px 0 0 0 alpha (@bg_highlight_color, 0.2),
        inset -1px 0 0 0 alpha (@bg_highlight_color, 0.2),
        inset 0 -1px 0 0 alpha (@bg_highlight_color, 0.3),
        0 0 0 1px alpha (#000, 0.05),
        0 1px 2px alpha (#000, 0.2);
}

dialog button label {
    text-shadow: none;
    font-weight: 500;
}

dialog button.suggested-action {
    background-color: @accent_color_100;
    color: @accent_color_900;
    box-shadow:
        inset 0 1px 0 0 alpha (@bg_highlight_color, 0.08),
        inset 1px 0 0 0 alpha (@bg_highlight_color, 0.02),
        inset -1px 0 0 0 alpha (@bg_highlight_color, 0.02),
        inset 0 -1px 0 0 alpha (@bg_highlight_color, 0.05),
        0 0 0 1px alpha (@accent_color_300, 0.4),
        0 1px 2px alpha (#000, 0.3);
}

dialog.destructive-action button.suggested-action {
    background-color: @error_color;
    color: #fff;
    box-shadow:
        inset 0 1px 0 0 alpha (@bg_highlight_color, 0.08),
        inset 1px 0 0 0 alpha (@bg_highlight_color, 0.02),
        inset -1px 0 0 0 alpha (@bg_highlight_color, 0.02),
        inset 0 -1px 0 0 alpha (@bg_highlight_color, 0.05),
        0 0 0 1px alpha (@STRAWBERRY_500, 0.8),
        0 1px 2px alpha (@STRAWBERRY_900, 0.8);
}

dialog.destructive-action .primary {
    color: @error_color;
}

dialog.destructive-action .dialog-action-box {
   // background: alpha (@STRAWBERRY_100, 0.2);
}
```

**Comment by micahilbery** [comment]
Ideas for dark suggested action buttons:
![dark-suggested-action-button](https://user-images.githubusercontent.com/10608836/76244306-cce3dd80-6207-11ea-91d6-a3c46d16c5c5.png)
![image11](https://user-images.githubusercontent.com/10608836/76244323-d53c1880-6207-11ea-94ed-6eb5576bbb94.png)
 Destructive action vs Red suggested action
![suggested vs destructive](https://user-images.githubusercontent.com/10608836/76244327-d705dc00-6207-11ea-9244-31acdcf6886c.png)

**Comment by cassidyjames** [comment]
@micahilbery It looks like the darkest accent color for suggested actions is darker than the normal button here, which almost makes the suggested button look like it's insensitive. I wonder if we can tint the normal button color toward the accent color for suggested actions, and if that would mean we'd need to either still use white text for contrast, or mix the accent text color with white to lighten it up?

**Comment by danirabbit** [comment]
Yeah I haven’t color checked those, but I agree they look low contrast and much darker than the regular button color.

Maybe we could consider a different approach for destructive actions altogether like a regular button with a red border or with a warning icon or some kind of other distinctive marker

**Comment by lainsce** [comment]
Maybe we could always pair a destructive action dialog with a warning icon, and never use a warning icon on a dialog that isn't leading to some destructive action, that way we could keep the soft red on a red accent system, because people would then get used to seeing warning paired with destructive actions.

**Comment by micahilbery** [comment]
@cassidyjames Here's it using the mix of the base color and the accent color for the background and a mix of the accent color and white for the text:
![Screenshot from 2020-03-09 14 43 22](https://user-images.githubusercontent.com/10608836/76251414-ef302800-6214-11ea-8b9d-3d907a5e9565.png)
![Screenshot from 2020-03-09 14 43 38](https://user-images.githubusercontent.com/10608836/76251471-04a55200-6215-11ea-80fa-b4b5e07f28b7.png)
![Screenshot from 2020-03-09 14 43 59](https://user-images.githubusercontent.com/10608836/76251417-efc8be80-6214-11ea-86f1-8673fd9fe62d.png)
![Screenshot from 2020-03-09 14 44 09](https://user-images.githubusercontent.com/10608836/76251418-efc8be80-6214-11ea-8f61-4c0d83ea95d2.png)
![Screenshot from 2020-03-09 14 44 20](https://user-images.githubusercontent.com/10608836/76251419-f0615500-6214-11ea-8c68-b5f54f9b23ac.png)
![Screenshot from 2020-03-09 14 44 30](https://user-images.githubusercontent.com/10608836/76251421-f0615500-6214-11ea-9043-5e6442ea48d1.png)

**Comment by jtrees** [comment]
Is it just me or does anybody else get bugged by that text alignment? Maybe something more like this?

![76241252-ed08a200-61f1-11ea-8cc2-0752a4b59f7d_5](https://user-images.githubusercontent.com/5802758/76267299-91c6c600-626a-11ea-972c-2a1367853592.png)

**Comment by danirabbit** [comment]
@jtrees this is unfortunately not possible because of the way the widgets are constructed. I've filed [this issue](https://github.com/elementary/stylesheet/issues/593) to at least try to align checkbuttons and expanders. But, in Gtk this is actually a single widget that includes the checkbox itself and its label. So the best we can do is:

![Screenshot from 2020-03-10 08 33 02@2x](https://user-images.githubusercontent.com/7277719/76329413-c7d76a80-62a9-11ea-9703-844a5d9156e2.png)

**Comment by jtrees** [comment]
@danrabbit Huh, that's a shame.

**Comment by hanaral** [comment]
@danrabbit Would it be possible to have a more prominent outline and sheen to the buttons? They look a little too soft compared to the more flat and contrasty style of the popup panel.

**Comment by hanaral** [comment]
I've been trying out the CSS on password popups, the bottom half of the dialogue looks much better when we copy the contrast of the text box 
![image](https://user-images.githubusercontent.com/58219504/79134057-fc58ad80-7da4-11ea-8305-e505ddec0bd1.png)
(I haven't tried it on a HiDPI screen, but it looks good on my non-retina macbook)

**Comment by CainCosta** [comment]
I really like the one with white text labels on the original post by Daniel. 

Material design achieves a nice balance with this by setting in-box text levels at 85-90% opacity. You still get high contrast while not looking completely out of place.

**Comment by hanaral** [comment]
@CainCosta Totally agree that the white text looks great, and it's akin to the already existing buttons too. I think the reason Daniel only gave the destructive buttons white text is so they can be distinguished from the user accented ones.
